File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        
        // Check if start or end cell is blocked
        if (grid[0][0] == 1 || grid[n-1][m-1] == 1) {
            return -1;
        }
        
        // Direction vectors for 8 possible moves
        vector<pair<int, int>> delta = {
            {1, 1}, {1, 0}, {1, -1}, {0, -1}, {0, 1},
            {-1, 1}, {-1, 0}, {-1, -1}
        };
        
        // Queue for BFS
        queue<pair<int, int>> q;
        q.push({0, 0});
        grid[0][0] = 2; // Mark the starting cell as visited
        int ans = 1;
        
        // BFS traversal
        while (!q.empty()) {
            int levelSize = q.size();
            for (int i = 0; i < levelSize; ++i) {
                int x = q.front().first;
                int y = q.front().second;
                q.pop();
                
                // Check if we reached the bottom-right corner
                if (x == n - 1 && y == m - 1) {
                    return ans;
                }
                
                // Explore all 8 possible directions
                for (const auto& d : delta) {
                    int xx = x + d.first;
                    int yy = y + d.second;
                    
                    if (xx == n - 1 && yy == m - 1) {
                        return ans + 1;
                    }
                    
                    // Check if the next cell is within grid bounds and not visited
                    if (xx >= 0 && xx < n && yy >= 0 && yy < m && grid[xx][yy] == 0) {
                        grid[xx][yy] = 2; // Mark the cell as visited
                        q.push({xx, yy});
                    }
                }
            }
            ++ans;
        }
        
        return -1; // Return -1 if no path found
    }
};

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (n == 0) return -1;
        int m = grid[0].length;
        
        if (grid[0][0] == 1 || grid[n-1][m-1] == 1) {
            return -1;
        }
        
        int[][] directions = {{1, 1}, {1, 0}, {1, -1}, {0, -1}, {0, 1}, {-1, 1}, {-1, 0}, {-1, -1}};
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[] {0, 0});
        grid[0][0] = 2;  // Mark as visited
        int pathLength = 0;
        
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            pathLength++;
            for (int i = 0; i < levelSize; i++) {
                int[] cell = queue.poll();
                int x = cell[0];
                int y = cell[1];
                
                if (x == n - 1 && y == m - 1) {
                    return pathLength;
                }
                
                for (int[] direction : directions) {
                    int xx = x + direction[0];
                    int yy = y + direction[1];
                    
                    if (xx == n - 1 && yy == m - 1) {
                        return pathLength + 1;
                    }
                    
                    if (inGrid(xx, yy, n, m) && grid[xx][yy] == 0) {
                        queue.offer(new int[] {xx, yy});
                        grid[xx][yy] = 2;  // Mark as visited
                    }
                }
            }
        }
        
        return -1;
    }
    
    private boolean inGrid(int x, int y, int n, int m) {
        return 0 <= x && x < n && 0 <= y && y < m;
    }
}
